fainder/execution/new_runner.py

```diff
@@ -55,6 +56,57 @@ def run_exact(
         id_filter=set(np.setdiff1d(recall_result, precision_result, assume_unique=True)),
     )
     logger.debug(f"profile-scan took {time.perf_counter() - pscan_start:.5f}s")
+
+    result = np.union1d(pscan_result, precision_result)
+
+    end = time.perf_counter()
+    return result, end - start
+
+
+def run_exact_parallel(
+    fainder_index: tuple[list[PctlIndex], list[NDArray[np.float64]]],
+    query: PctlQuery,
+    parallel_processor: ParallelHistogramProcessor,
+    id_filter: ArrayLike | None = None,
+) -> tuple[NDArray[np.uint32], float]:
+    """Run an exact percentile query using parallel processing.
+
+    This function is thread-safe and can be called from multiple threads simultaneously.
+
+    Args:
+        fainder_index: The index to query
+        query: The percentile query to run
+        parallel_processor: The initialized ParallelHistogramProcessor instance
+
+    Returns:
+        A tuple of (result array, runtime in seconds)
+    """
+
+    start = time.perf_counter()
+
+    # Stage 1: Get recall results
+    recall_result = query_index_single(query, *fainder_index, index_mode="recall")
+
+    # Stage 2: Get precision results
+    precision_result = query_index_single(query, *fainder_index, index_mode="precision")
+
+    # Stage 3: Process histograms in parallel for the candidates
+    pscan_start = time.perf_counter()
+    candidates = np.setdiff1d(recall_result, precision_result)
+
+    if id_filter is not None:
+        candidates = np.intersect1d(candidates, id_filter, assume_unique=True)
+    if candidates.size <= 0:
+        pscan_result = np.array([], dtype=np.uint32)
+    else:
+        pscan_result = parallel_processor.query(query, id_filter=candidates)
+
+    logger.debug(f"Parallel profile-scan took {time.perf_counter() - pscan_start:.5f}s")
+
+    # Combine results
+    if id_filter is not None:
+        precision_result = np.intersect1d(precision_result, id_filter, assume_unique=True)
+
```

Comment:
Move this directly under line 91 and add a comment stating that we do this because passing the filter to query_index_single is slower.